# VarGenius - Introduction

*VarGenius* is a pipeline which allows to call and annotate variants. It is based on the GATK best practices and uses the most common programs for quality control (FastQC), trimming of raw reads (TrimGalore), alignment (BWA) and for the management of the alignment results (PICARD and SAMTOOLS). While Annovar is used for variants annotation. 
*VarGenius* creates a database that is used to store information about samples and analyses and about the detected variants and their annotations.
The database allows to both manage the analyses in an efficient way and to store the variants and their annotations for further downstream analyses.

## Analysis steps

*VarGenius* has been written for the Telethon Undiagnosed Disease Project (TUDP) whose protocol follows strictly the GATK best practices for the variants detection. 

The overall set of steps implemented is the following: quality check, trimming, merge paired ends, alignment against reference, mark and removed duplicates, view, sort and index bam file, creation of a target for the realignment, indel realignment, base recalibrator, print recalibrated reads, merge sam files, mark and remove duplicates, variant calling, cat variants, variants recalibration (SNPs and INDELs), apply variant recalibration, genotyping, variant filtering, phasing, annotation, output production.

Each step uses different software from the following toolkits: Samtools, Picard tools, GATK, BWA-kit, TrimGalore, FastQC. The **steps** are merged into **tasks** that can be called using the command line: quality check, alignment, refinement, variant calling, genotype, variant filtering, phasing, output production and statistics.

Each task can be executed independently from each other but almost all steps need a specific previous step to be successfully executed.
The database contains flags for each read file, sample or group describing which step has been performed. Hence, when a module is called the program checks if the step needed has been already executed.

Letâ€™s look more in detail into the different execution modules:

*Quality Check* (quality check, trimming): optionally redefines the quality metric of the reads and allows to visually check the quality of the reads (FASTQC) and permits to remove the adapter and trim the reads for quality less than a given score (TrimGalore).

*Alignment against reference* (merge paired ends, alignment against reference, mark and removed duplicates, view, sort and index bam file): allows to align the reads against the given reference transcriptome (BWA-MEM), to evaluate the ALT contigs and other outliers of the genome (BWA-KIT), makes the grouping of reads (AddOrRemoveReadGroups or BWA-KIT) and sorts the BAM files (samtools).

*Alignment refinement* (creation of a target for the realignment, indel realignment, base recalibrator, print recalibrated reads): this is the first step performed with GATK exploiting their Best Practices. *VarGenius* implements the step of Indel realignment (RealignerTargetCreator and IndelRealigner) and the step of base quality recalibration (BaseRecalibrator and PrintReads).

This step uses known variants to do the job. GATK provides an up-to-date resource bundle with all the needed datasets here: ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle.

*Merge read files* ( merge sam files, re-mark duplicates): optionally used if more read files per sample are present.

*Variant Calling* ( variant calling, cat variants): Variant discovery is performed using GATK HaplotypeCaller and is performed per-sample. After that the variants are called the genotyping step is performed.

*Variant Quality Score Recalibration* (variants recalibration (SNPs and INDELs), apply variant recalibration): which permits the recalibration of variants based on the use of known variants.

*Genotyping*: is the joint genotype which can be used whenever you want to analyze different samples together. You can use this step both to discover variants with low coverage from multiple analyses of single samples and to make family analyses (Trio, Quartet, etc).

As from the GATK Best Practices, the only part that really needs to be done jointly and cannot be separated in different processes is the Genotyping step. Since GATK3.0 variant calling can be used individually per-sample in GVCF mode followed by a joint genotyping step.
Please refer to the [GATK forum](http://gatkforums.broadinstitute.org/gatk/discussion/3670/the-benefit-of-using-a-cohort) for more information:

*Variant Filtration*: which permits a first level of filtering for variants which have low quality

*Phasing*: gets the exact haplotype using parents information. The phasing is executed using PhaseByTransmission and is an optional step which is called only if you have the trio. A PED file is automatically generated by *VarGenius* if the sample sheet contains the kinship information.

*Annotate, import and get output*

The aim of the final output task is to both annotate the variants and to generate the final tabular output with statistics.
To obtain a faster process we divided the output execution in two steps. The first step involves the annotation of the variants and the output creation while a parallel step is the insertion of variants into the database.
The variants are annotated using Annovar.



Now that you know what is the purpose of *VarGenius*, go and install it! [INSTALL](https://github.com/frankMusacchia/VarGenius/blob/master/GUIDE/INSTALL.md)

---------------------------------

If you get some error during the installation or the running of *VarGenius* please see the FAQ page of the user guide or ask on the google group: https://groups.google.com/forum/#!forum/vargenius
